ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,192 @@
+                blockedTime = Math.min(threadTimes, blockedTime);
+            }
+        }
+        final double notBlocked = windowSize - blockedTime;
+        lastSampleTime = sampleTime;
+        logger.info("The current processing ratio is " + percentage(notBlocked, windowSize) + "%");
+        return percentage(notBlocked, windowSize);
+    }
+
+    // public for testing
+    public double getProcessingRatio(final String threadName, final KafkaStreams streams, final long windowStart, final double windowSize) {
+        try {
+            final Collection<? extends Metric> allMetrics = streams.metrics().values();
+            final Map<String, Double> consumerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(CONSUMER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && !m.metricName().tags().getOrDefault("client-id", "").contains("restore-consumer")
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> restoreConsumerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(CONSUMER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && m.metricName().tags().getOrDefault("client-id", "").contains("restore-consumer")
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> producerMetrics = allMetrics.stream()
+                .filter(m -> m.metricName().group().equals(PRODUCER_METRICS)
+                    && metrics.contains(m.metricName().name())
+                    && m.metricName().tags().getOrDefault("client-id", "").contains(threadName))
+                .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final Map<String, Double> threadMetrics = allMetrics.stream()
+                    .filter(m -> m.metricName().group().equals(STREAM_METRICS) &&
+                            m.metricName().tags().getOrDefault(THREAD_ID, "").equals(threadName) &&
+                            metrics.contains(m.metricName().name()))
+                    .collect(Collectors.toMap(k -> k.metricName().name(), v -> (Double) v.metricValue()));
+
+            final List<Metric> startTimeList = allMetrics.stream()
+                    .filter(m -> m.metricName().group().equals(STREAM_METRICS) &&
+                            m.metricName().tags().getOrDefault(THREAD_ID, "").equals(threadName) &&
+                            m.metricName().name().equals("thread-start-time"))
```

Comment:
Do we have this metric in KS really?

Reply:
I added that: https://github.com/apache/kafka/commit/b96f99c595694c954bec59d2ae1fdd9f66d249be but haven't opened the KIP to add to streams

We haven't productionized any of the streams changes but we probably want to soon since the processing ratio is somewhat useless without the thread start time. I think we should change the default time for thread start to be the last poll time to give us a slightly more accurate reading. WDYT @rodesai ?